codalab/lib/bundle_cli.py

```diff
@@ -1170,26 +1185,134 @@ def derive_bundle(self, bundle_type, command, targets, metadata):
     def do_run_command(self, args):
         client, worksheet_uuid = self.parse_client_worksheet_uuid(args.worksheet_spec)
         args.target_spec, args.command = cli_util.desugar_command(args.target_spec, args.command)
-        targets = self.parse_key_targets(client, worksheet_uuid, args.target_spec)
         metadata = self.get_missing_metadata(RunBundle, args)
-        new_bundle = client.create(
-            'bundles',
-            self.derive_bundle(RunBundle.BUNDLE_TYPE, args.command, targets, metadata),
-            params={'worksheet': worksheet_uuid},
+
+        if args.local:
+            docker_image = metadata.get('request_docker_image', None)
+            if docker_image is None:
+                raise UsageError('--request-docker-image [docker-image] must be specified when running in local mode')
+
+            uuid = generate_uuid()
+            bundle_path = '/opt/{}'.format(uuid)
+            command = args.command
+            request_network = None
+            dependencies = [
+                (u'{}'.format(key), u'/{}_dependencies/{}'.format(uuid, key)) for key, target in self.parse_target_specs(args.target_spec)
+            ]
+
+            # Set up a directory to store the bundle.
+            remove_path(bundle_path)
+            os.mkdir(bundle_path)
+
+            for dependency_path, docker_dependency_path in dependencies:
+                child_path = os.path.join(bundle_path, dependency_path)
+                os.symlink(docker_dependency_path, child_path)
+
+            dc = DockerClient()
+            container_id = dc.start_container(bundle_path, uuid, command, docker_image, request_network, dependencies)
+            print >>self.stdout, '===='
+            print >>self.stdout, 'ContainerID: ', container_id
+            print >>self.stdout, 'Local Bundle ID: ', uuid
+            print >>self.stdout, 'You can find local bundle contents in: ', bundle_path
+            print >>self.stdout, '===='
+        else:
+            targets = self.parse_key_targets(client, worksheet_uuid, args.target_spec)
+            new_bundle = client.create(
+                'bundles',
+                self.derive_bundle(RunBundle.BUNDLE_TYPE, args.command, targets, metadata),
+                params={'worksheet': worksheet_uuid},
+            )
+
+            print >>self.stdout, new_bundle['uuid']
+            self.wait(client, args, new_bundle['uuid'])
+
+    @Commands.command(
+        'edit-image',
+        help='Drop to an interactive container within an image as a bash shell to allow edits.',
+        arguments=(
+            Commands.Argument('target_spec', help=ALIASED_TARGET_SPEC_FORMAT, nargs='*', completer=TargetsCompleter),
+            Commands.Argument('--request-docker-image', help='The docker image to edit', required=True),
         )
+    )
+    def do_edit_image_command(self, args):
+        docker_image = args.request_docker_image
+
+        uuid = generate_uuid()
+        bundle_path = '/opt/{}'.format(uuid)
```

Comment:
Refactor so these paths only appear once (both for `run-local` and this function).